04-data-structures-part1.Rmd

```diff
@@ -2,336 +2,271 @@
+we see that that's a vector, too - *the columns of data we load into R data.frames are all vectors*, and that's the root of why R forces everything in a column to be the same basic data type.
+
+> ## Discussion 1 {.challenge}
 >
-> When you combine numbers using the concatenate function, `c()` the type
-> will automatically become "numeric", that is real/decimal numbers. If you
-> specifically want to create a vector of integers (whole numbers only),
-> you need to append each number with an L, i.e. `c(10L, 12L, 45L, 33L)`.
+> Why is R so opinionated about what we put in our columns of data?
+> How does this help us?
 >
 
-You can also use the concatenate function to add elements to a vector:
+You can also make vectors with explicit contents with the concatenate function:
 
 ```{r}
-x <- c(x, 57)
+x <- c(2,6,3)
 x
 ```
 
-> ## Challenge 2 {.challenge}
->
-> Vectors can only contain one atomic type. If you try to combine different
-> types, R will create a vector that is the least common denominator: the
-> type that is easiest to coerce to.
->
-> **Guess what the following do without running them first:**
->
-> ```{r, eval=FALSE}
-> xx <- c(1.7, "a")
-> xx <- c(TRUE, 2)
-> xx <- c("a", TRUE)
-> ```
->
-
-This is called implicit coercion.
-
-The coercion rule goes `logical` -> `integer` -> `numeric` -> `complex` ->
-`character`.
-
-You can also coerce vectors explicitly using the `as.<class_name>`. Example
+Given what we've learned so far, what do you think the following will produce?
 
 ```{r}
-as.numeric()
-as.character()
+x <- c(2,6,'3')
 ```
 
-R will try to do whatever makes the most sense for that value:
+This is something called *type coercion*, and it is the source of many surprises and the reason why we need to be aware of the basic data types and how R will interpret them. Consider:
 
 ```{r}
-as.character(x)
+x <- c('a', TRUE)
+x
+x <- c(0, TRUE)
+x
 ```
 
-```{r}
-as.complex(x)
-```
+The coercion rules go: `logical` -> `integer` -> `numeric` -> `complex` -> `character`. You can try to force coercion against this flow using the `as.` functions:
 
 ```{r}
-x <- 0:6
-as.logical(x)
+x <- c('0','2','4')
+x
+y <- as.numeric(x)
+y
+z <- as.logical(y)
+z
 ```
 
-This is behaviour you will find in many programming languages. 0 is
-FALSE, while every other number is treated as TRUE.
-Sometimes coercions, especially nonsensical ones won't work.
+As you can see, some surprising things can happen when R forces one basic data type into another! Nitty-gritty of type coercion aside, the point is: if your data doesn't look like what you thought it was going to look like, type coercion may well be to blame; make sure everything is the same type in your vectors and your columns of data.frames, or you will get nasty surprises!
 
-In some cases, R won't be able to do anything sensible:
+Concatenate will also append things to an existing vector:
 
 ```{r}
-x <- c("a", "b", "c")
-as.numeric(x)
-as.logical(x)
+x <- c('a', 'b')
+x
+x <- c(x, 'SWC')
+x
```

Comment:
Is it worth adding a callout here to the for-loop section? Just to point out that while appending things to vectors in this way will work, it is a bad practice. (because it slows R down so much)

Additionally, I think it's tricky in a classroom setting to overwrite `x` like this.  Better to suggest instructors do 
````r
x <- c(x, 'SWC')
```

Or even give the vectors memorable names, like `first_two_letters` and `first_two_swc` or whatever